execution_engine/src/core/runtime/host_function_flag.rs

```diff
@@ -0,0 +1,95 @@
+use std::{cell::Cell, rc::Rc};
+
+use tracing::error;
+
+/// A flag to indicate whether the current runtime call is made within the scope of a host function.
+///
+/// The flag is backed by an `Rc<Cell<u64>>`, meaning that clones will all share state.
+#[derive(Default, Clone)]
+pub(super) struct HostFunctionFlag {
+    /// A counter which, if non-zero, indicates that the `HostFunctionFlag` is `true`.
+    counter: Rc<Cell<u64>>,
+}
+
+impl HostFunctionFlag {
+    /// Returns `true` if this `HostFunctionFlag` has entered any number of host function scopes
+    /// without having exited them all.
+    pub(super) fn is_in_host_function_scope(&self) -> bool {
+        self.counter.get() != 0
+    }
+
+    /// Must be called when entering a host function scope.
+    ///
+    /// The returned `ScopedHostFunctionFlag` must be kept alive for the duration of the host
+    /// function call.  While at least one such `ScopedHostFunctionFlag` exists,
+    /// `is_in_host_function_scope()` returns `true`.
+    #[must_use]
+    pub(super) fn enter_host_function_scope(&self) -> ScopedHostFunctionFlag {
+        let new_count = self.counter.get().checked_add(1).unwrap_or_else(|| {
```

Comment:
I believe we should panic here, as the counter reaching `u64::MAX` is a programming error.

Reply:
I don't believe panicking is the best option, since that could ultimately cause a network crash.

I'm fairly convinced we won't be able to hit this code.  Just running something like
```rust
let counter = std::rc::Rc::new(std::cell::Cell::new(0_u64));
loop {
    let new_count = counter.get().checked_add(1).unwrap();
    counter.set(new_count);
}
```
would take over 100 years to execute before it panicked on my machine.

Reply:
Indeed, it is unlikely we will hit this code, but if we end up hitting it, it is a logic mistake because we would need a large chain of wasm to host calls. The rust book recommends panicking on programming errors: <https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html#guidelines-for-error-handling>.

In the unlikely scenario that we do hit it, aborting execution would be preferable to having an incorrect value in the counter (inconsistent state).